payment-settler/bin/payment-settler.js

```diff
@@ -0,0 +1,134 @@
+/** @import {MessageBatch} from 'cloudflare:workers' */
+import { getChainClient as defaultGetChainClient } from '../lib/chain.js'
+import { abi } from '../lib/filbeam.js'
+import { getDataSetsForSettlement } from '../lib/rail-settlement.js'
+import { TransactionMonitorWorkflow } from '@filbeam/workflows'
+import { handleTransactionRetryQueueMessage as defaultHandleTransactionRetryQueueMessage } from '../lib/queue-handlers.js'
+
+/**
+ * @typedef {{
+ *   type: 'transaction-retry'
+ *   transactionHash: string
+ * }} TransactionRetryMessage
+ */
+
+/**
+ * @typedef {{
+ *   ENVIRONMENT: 'dev' | 'calibration' | 'mainnet'
+ *   RPC_URL: string
+ *   FILBEAM_CONTRACT_ADDRESS: string
+ *   FILBEAM_CONTROLLER_ADDRESS_PRIVATE_KEY: string
+ *   DB: D1Database
+ *   TRANSACTION_MONITOR_WORKFLOW: import('cloudflare:workers').WorkflowEntrypoint
+ *   TRANSACTION_QUEUE: import('cloudflare:workers').Queue<TransactionRetryMessage>
+ * }} RailSettlementEnv
+ */
+
+export default {
+  /**
+   * @param {any} _controller
+   * @param {RailSettlementEnv} env
+   * @param {ExecutionContext} _ctx
+   * @param {{ getChainClient?: Function }} [options]
+   */
+  async scheduled(
+    _controller,
+    env,
+    _ctx,
+    { getChainClient = defaultGetChainClient } = {},
+  ) {
+    console.log('Starting rail settlement worker')
+
+    try {
+      const { publicClient, walletClient, account } = getChainClient(env)
+
+      const dataSetIds = await getDataSetsForSettlement(env.DB)
+
+      if (dataSetIds.length === 0) {
+        console.log('No active data sets found for settlement')
+        return
+      }
+
+      console.log(
+        `Found ${dataSetIds.length} data sets for settlement:`,
+        dataSetIds,
+      )
+
+      const { request } = await publicClient.simulateContract({
+        account,
+        abi,
+        address: env.FILBEAM_CONTRACT_ADDRESS,
+        functionName: 'settleCDNPaymentRailBatch',
+        args: dataSetIds.map((id) => BigInt(id)),
```

Comment:
We will eventually need to split this TX into multiple smaller ones, when the list of data sets becomes too large.

I think our situation is simpler in the payment-settler worker, because we don't have to settle all data sets at once. We can settle up to X data sets per worker invocation, where X is chosen so that we don't exceed the max calldata size. In the next run, the worker will pick next X data sets that haven't been settled yet, and so on.

Anyhow, we don't need that yet.

Please create a follow-up issue and add a code comment here to remind us about the limitation.

Reply:
[Follow-up issue](https://github.com/filbeam/worker/issues/353)